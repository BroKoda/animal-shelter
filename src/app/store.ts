import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit'
import homeReducer from '../pages/Home/HomeSlice'
import loginReducer from '../pages/Login/LoginSlice'
import registrationReducer from '../pages/Registration/RegistrationSlice'
import residentsReducer from '../pages/Residents/ResidentsSlice'
import newsReducer from '../pages/News/NewsSlice'
import contactReducer from '../pages/Contact/ContactSlice'
import dashboardReducer from '../pages/Dashboard/DashboardSlice'

export const store = configureStore({
  reducer: {
    home: homeReducer,
    news: newsReducer,
    login: loginReducer,
    contact: contactReducer,
    dashboard: dashboardReducer,
    residents: residentsReducer,
    registration: registrationReducer,
  }
})

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
