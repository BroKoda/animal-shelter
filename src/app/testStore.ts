import homeReducer from '../pages/Home/HomeSlice'
import loginReducer from '../pages/Login/LoginSlice'
import registrationReducer from '../pages/Registration/RegistrationSlice'
import residentsReducer from '../pages/Residents/ResidentsSlice'
import newsReducer from '../pages/News/NewsSlice'
import contactReducer from '../pages/Contact/ContactSlice'
import { combineReducers, configureStore, PreloadedState } from '@reduxjs/toolkit'

const testRootReducer = combineReducers({
  home: homeReducer,
  news: newsReducer,
  login: loginReducer,
  contact: contactReducer,
  residents: residentsReducer,
  registration: registrationReducer
})

export function setupStore(preloadedState?: PreloadedState<TestRootReducer>): any {
  return configureStore({
    reducer: testRootReducer,
    preloadedState
  })
}

export type TestAppStore = ReturnType<typeof setupStore>
export type TestRootReducer = ReturnType<typeof testRootReducer>