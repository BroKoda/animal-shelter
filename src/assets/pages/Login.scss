@import '../variables-mixins';

.login-container {
  height: 100vh;
  position: relative;

  .login-box {
    background-color: $gray;
    border-radius: 20px;
    height: 100%;

    .call-to-action-button {
      width: 100%;
    }

    .secondary-button {
      width: 100%;
    }
  }
}