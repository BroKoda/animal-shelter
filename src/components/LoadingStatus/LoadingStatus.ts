export const enum LoadingStatus {
  initial = 'initial',
  complete = 'complete',
  loading = 'loading',
  error = 'error'
}