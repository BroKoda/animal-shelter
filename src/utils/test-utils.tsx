import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import type { PreloadedState } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import { setupStore } from '../app/testStore'
import type { TestAppStore, TestRootReducer } from '../app/testStore'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<TestRootReducer>
  store?: TestAppStore
}

export function renderWithProviders(
  ui: JSX.Element,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
): any {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}