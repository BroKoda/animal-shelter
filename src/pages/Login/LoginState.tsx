import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'

export interface LoginState {
  userToLogin: LoginDetails
  loggedInUser?: any
  error?: string
  status: LoadingStatus
}

export interface LoginDetails {
  email?: string
  password?: string
}