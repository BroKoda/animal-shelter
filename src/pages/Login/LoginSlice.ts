import { LoginDetails, LoginState, User } from './LoginState'
import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'
import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { RootState } from '../../app/store'
import { getAuth, signInWithEmailAndPassword, signOut } from 'firebase/auth'
import { collection, getDocs, query, where } from 'firebase/firestore'
import { db } from '../../firebase'

const initialState: LoginState = {
  userToLogin: {},
  status: LoadingStatus.initial,
  userStatus: LoadingStatus.initial,
  signOutUserStatus: LoadingStatus.initial
}

export const logInUser = createAsyncThunk('logInUser', async ({email, password}: LoginDetails) => {
  if (email != null && password != null) {
    await signInWithEmailAndPassword(getAuth(), email, password)
    return true
  }
})

export const getUserDetails = createAsyncThunk('getUserDetails', async (email: string | null) => {
  if (email != null) {
    const userData = await getDocs(query(collection(db, 'users'), where('email', '==', email)))
    return userData.docs[0].data() as User
  }
})

export const signOutUser = createAsyncThunk('signOutUser', async () => {
  const auth = getAuth()
  await signOut(auth)
  return true
})

const loginSlice = createSlice({
  name: 'login',
  initialState,
  reducers: {
    resetLogin: () => initialState,
    setLoginEmail: (state: LoginState, action: PayloadAction<string>) => {
      state.userToLogin.email = action.payload
    },
    setLoginPassword: (state: LoginState, action: PayloadAction<string>) => {
      state.userToLogin.password = action.payload
    },
  },
  extraReducers: builder =>  {
    builder
      .addCase(logInUser.pending, (state) => {
        state.status = LoadingStatus.loading
      })
      .addCase(logInUser.fulfilled, (state) => {
        state.status = LoadingStatus.complete
      })
      .addCase(logInUser.rejected, (state) => {
        state.status = LoadingStatus.error
        state.error = 'Nem megfelelő email cím vagy jelszó!'
      })
      .addCase(getUserDetails.pending, (state) => {
        state.userStatus = LoadingStatus.loading
      })
      .addCase(getUserDetails.fulfilled, (state, action) => {
        state.userStatus = LoadingStatus.complete
        state.user = action.payload
      })
      .addCase(getUserDetails.rejected, (state) => {
        state.userStatus = LoadingStatus.error
      })
      .addCase(signOutUser.pending, (state) => {
        state.signOutUserStatus = LoadingStatus.loading
      })
      .addCase(signOutUser.fulfilled, (state) => {
        state.signOutUserStatus = LoadingStatus.complete
      })
      .addCase(signOutUser.rejected, (state) => {
        state.signOutUserStatus = LoadingStatus.error
      })
  }
})

export const { resetLogin, setLoginEmail, setLoginPassword } = loginSlice.actions
export const selectLogin = (state: RootState) => state.login
export default loginSlice.reducer
