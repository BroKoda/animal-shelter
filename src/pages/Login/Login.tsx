import React, { ChangeEvent, useCallback, useEffect } from 'react'
import { Link, useNavigate } from 'react-router-dom'
import '../../assets/pages/Login.scss'
import { useAppDispatch, useAppSelector } from '../../app/hooks'
import { logInUser, resetLogin, setLoginEmail, setLoginPassword } from './LoginSlice'
import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'

const Login = (): JSX.Element => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  const { status, error } = useAppSelector((state) => state.login)
  const { email, password } = useAppSelector((state) => state.login.userToLogin)

  useEffect(() => {
    if (status === LoadingStatus.complete) {
      dispatch(resetLogin())
      navigate('/')
    }
  }, [dispatch, status])

  const handleLogin = useCallback(() => {
    void dispatch(logInUser({email, password}))
  }, [dispatch, email, password])

  const handleSetLoginEmail = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setLoginEmail(event.target.value))
  }, [dispatch, email])

  const handleSetLoginPassword = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setLoginPassword(event.target.value))
  }, [dispatch, password])

  return (
    <div className="container login-container">
      <div className="row justify-content-center h-100">
        <div className="col-12 col-md-8 col-lg-6 col-xl-5 align-self-center">
          <div className="container login-box py-4 px-4">
            <div className="row">
              <div className="col-12">
                <h2>Belépés</h2>
                <p className='login-error'>{error}</p>
                <form className='my-4'>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-at"></i>
                    </div>
                    <input
                      type="email"
                      className="form-control"
                      placeholder='Email'
                      id="loginEmail"
                      value={email ?? ''}
                      onChange={handleSetLoginEmail}
                    />
                  </div>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-lock"></i>
                    </div>
                    <input
                      type="password"
                      className="form-control"
                      placeholder='Jelszó'
                      id="loginPassword"
                      value={password ?? ''}
                      onChange={handleSetLoginPassword}
                    />
                  </div>
                </form>
              </div>
            </div>
            <div className='row'>
              <div className='col-12 mb-3 text-end redirect-to-registration'>
                <Link to={'/regisztracio'}>
                  <span>Nincs még felhasználód? Regisztrálj!</span>
                </Link>
              </div>
            </div>
            <div className='row'>
              <div className='col-6'>
                <Link to={'/'}>
                  <button className='button secondary-button'>
                    Vissza
                  </button>
                </Link>
              </div>
              <div className='col-6'>
                <button className='button call-to-action-button' onClick={handleLogin}>
                  Belépés
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

export default Login
