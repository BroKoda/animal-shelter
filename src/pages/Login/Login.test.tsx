import { LoginState } from './LoginState'
import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'
import LoginSlice, {
  getUserDetails,
  logInUser,
  selectLogin,
  setLoginEmail,
  setLoginPassword,
  signOutUser
} from './LoginSlice'
import { renderWithProviders } from '../../utils/test-utils'
import { BrowserRouter } from 'react-router-dom'
import Login from './Login'
import { RootState, store } from '../../app/store'

describe('Login page tests', () => {
  const appState = store.getState()
  const initialState: LoginState = {
    userToLogin: {},
    status: LoadingStatus.initial,
    userStatus: LoadingStatus.initial,
    signOutUserStatus: LoadingStatus.initial
  }

  it('Should set login email', () => {
    expect(LoginSlice(initialState, setLoginEmail('test@email.com'))).toEqual({
      userToLogin: {
        email: 'test@email.com'
      },
      status: LoadingStatus.initial,
      userStatus: LoadingStatus.initial,
      signOutUserStatus: LoadingStatus.initial
    })
  })

  it('Should set login password', () => {
    expect(LoginSlice(initialState, setLoginPassword('testPass'))).toEqual({
      userToLogin: {
        password: 'testPass'
      },
      status: LoadingStatus.initial,
      userStatus: LoadingStatus.initial,
      signOutUserStatus: LoadingStatus.initial
    })
  })

  it('Should render login page', () => {
    const {container} = renderWithProviders(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('Should set loading state when logging in user', (done) => {
    const nextState: LoginState = LoginSlice(initialState, logInUser.pending)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).status).toEqual(LoadingStatus.loading)
    done()
  })

  it('Should set loading state when logging in user', (done) => {
    const nextState: LoginState = LoginSlice(initialState, logInUser.fulfilled)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).status).toEqual(LoadingStatus.complete)
    done()
  })

  it('Should set loading state when logging in user', (done) => {
    const nextState: LoginState = LoginSlice(initialState, logInUser.rejected)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).status).toEqual(LoadingStatus.error)
    done()
  })

  it('Should set loading state when fetching user data', (done) => {
    const nextState: LoginState = LoginSlice(initialState, getUserDetails.pending)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).userStatus).toEqual(LoadingStatus.loading)
    done()
  })

  it('Should set loading state when fetching user data', (done) => {
    const nextState: LoginState = LoginSlice(initialState, getUserDetails.fulfilled)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).userStatus).toEqual(LoadingStatus.complete)
    done()
  })

  it('Should set loading state when fetching user data', (done) => {
    const nextState: LoginState = LoginSlice(initialState, getUserDetails.rejected)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).userStatus).toEqual(LoadingStatus.error)
    done()
  })

  it('Should set loading state when logging out user', (done) => {
    const nextState: LoginState = LoginSlice(initialState, signOutUser.pending)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).signOutUserStatus).toEqual(LoadingStatus.loading)
    done()
  })

  it('Should set loading state when logging out user', (done) => {
    const nextState: LoginState = LoginSlice(initialState, signOutUser.fulfilled)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).signOutUserStatus).toEqual(LoadingStatus.complete)
    done()
  })

  it('Should set loading state when logging out user', (done) => {
    const nextState: LoginState = LoginSlice(initialState, signOutUser.rejected)
    const rootState: RootState = {... appState, login: nextState}

    expect(selectLogin(rootState).signOutUserStatus).toEqual(LoadingStatus.error)
    done()
  })
})