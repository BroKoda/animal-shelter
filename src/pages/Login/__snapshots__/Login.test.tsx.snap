// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page tests Should render login page 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row justify-content-center h-100"
    >
      <div
        class="col-12 col-md-8 col-lg-6 col-xl-5 align-self-center"
      >
        <div
          class="container login-box py-4 px-4"
        >
          <div
            class="row"
          >
            <div
              class="col-12"
            >
              <h2>
                Belépés
              </h2>
              <p
                class="login-error"
              />
              <form
                class="my-4"
              >
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-at"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Email
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="loginEmail"
                    placeholder="Email"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-lock"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Jelszó
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="loginPassword"
                    placeholder="Jelszó"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
              </form>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-12 mb-3 text-end redirect-to-registration"
            >
              <a
                href="/regisztracio"
              >
                <span>
                  Nincs még felhasználód? Regisztrálj!
                </span>
              </a>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-6"
            >
              <a
                href="/"
              >
                <button
                  class="button secondary-button"
                >
                  Vissza
                </button>
              </a>
            </div>
            <div
              class="col-6"
            >
              <button
                class="button call-to-action-button"
              >
                Belépés
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
